File: 0410-split-array-largest-sum/0410-split-array-largest-sum.py
class Solution:
    def isPossible(self, nums: List[int], k: int, mid: int) -> bool:
        subArrSum = 0
        numberOfElements = 1

        for x in nums:
            # If single element is bigger than mid, splitting not possible
            if x > mid:
                return False

            if subArrSum + x <= mid:
                subArrSum += x
            else:
                numberOfElements += 1
                if numberOfElements > k:
                    return False
                subArrSum = x  # FIXED: lowercase correct variable name
        
        return True

    def splitArray(self, nums: List[int], k: int) -> int:
        s = max(nums)  # optimization
        e = sum(nums)
        ans = -1

        while s <= e:
            mid = s + (e - s) // 2
            if self.isPossible(nums, k, mid):  # FIXED: call properly
                ans = mid
                e = mid - 1
            else:
                s = mid + 1
        return ans

File: 0410-split-array-largest-sum/0410-split-array-largest-sum.cpp
class Solution {
private:
    bool isPossible(vector<int>& nums, int k, int mid){
        int SumSub = 0;
        int numberofelements = 1;

        for (int i:nums){
            if (i > mid){
                return false;
            }
            if(SumSub + i <= mid){
                SumSub+=i;
            }
            else{
                numberofelements+=1;
                if (numberofelements>k || SumSub > mid){
                    return false;
                }
                SumSub = i;
                
            }
        }
        return true;


    }
public:
    int splitArray(vector<int>& nums, int k) {
        int tsum = 0;
        int max_arr = 0;
        for(int i : nums){
            tsum += i;
            max_arr = max(max_arr,i);
        }
        int start = max_arr;
        int end = tsum;
        int ans = -1;
        while (start <= end){
            int mid = start + (end-start)/2;
            if (isPossible(nums,k,mid)){
                ans = mid;
                end = mid-1;
            }
            else{
                start = mid + 1;
            }
        }
        return ans;

        
    }
};